facts - fix incorrect time for some date_time_facts

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
The `iso8601_micro` and `iso8601` facts incorrectly called `now.utcnow()`, resulting in a new timestamp at the time it was called, not a conversion of the previously stored timestamp.

Correct this by capturing the epoch timestamp once then creating local and UTC `datetime` objects using this epoch value.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/facts/system/date_time.py`
##### ADDITIONAL INFORMATION
Follow up from #70449

- [x] Add unit tests for various time zones and DST states

samdoran committed:
facts - fix incorrect time for some date_time_facts

The iso8601_micro and iso8601 facts incorrectly called now.utcnow(), resulting
in a new timestamp at the time it was called, not a conversion of the previously
stored timestamp.

Correct this by capturing the UTC timestamp once then calculating the local
time using the UTC offset of the current system.

samdoran commented:
@nickgryg I would appreciate if you could take a look.

sivel commented on lib/ansible/module_utils/facts/system/date_time.py:
```diff
@@ -32,7 +36,10 @@ def collect(self, module=None, collected_facts=None):
 
-        now = datetime.datetime.now()
+        utcnow = datetime.utcnow()
+        offset_delta = timedelta(seconds=time.timezone)
```
I might recommend using `time.time()` as the base, and then calculate local and utc using:

```
ts = time.time()
now = datetime.datetime.fromtimestamp(ts)
utcnow = datetime.datetime.utcfromtimestamp(ts)
```

samdoran replied:
That’s a great suggestion. Thanks! I went through several different approaches (and spent way too much time reading docs) but I didn’t try using `time.time()`. I’ll give it a shot.

webknjaz replied:
Why does `datetime.utcnow()` not work here?

sivel replied:
The problem is getting both utc and local time, that match.  Calling both gives you different datetime objects. To convert between utc and local, you have to know the offset, potentially including DST to calculate it.  Going the `time.time` route, and creating a datetime from the timestamp resolves that complexity.

webknjaz replied:
Ah, right, forgot that it's PITA in Python 2... Python 3 is easier:
```python
$ ipython
Python 3.7.1 (default, Jan 28 2019, 08:25:13) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from datetime import datetime

In [2]: utcn = datetime.utcnow()

In [3]: utcn
Out[3]: datetime.datetime(2020, 7, 16, 21, 43, 35, 56745)

In [4]: utcn.astimezone()
Out[4]: datetime.datetime(2020, 7, 16, 21, 43, 35, 56745, tzinfo=datetime.time
zone(datetime.timedelta(seconds=7200), 'CEST'))
```

samdoran replied:
After digging into this far too much, I think I'm going to change the testing approach to validate the calls to `datetime.datetime.strftime()` since mocking out `time` is far from trivial.

nickgryg commented:
I like @sivel 's suggestion.

But what if we just change less code here? We need to properly get UTC datetime object from NOW (let me call it like that) datetime one. So, what's about:
```
In [1]: import datetime

In [2]: now = datetime.datetime.now()

In [3]: utc_now = datetime.datetime.utcfromtimestamp(int(now.strftime('%s'))) + datetime.timedelta(microseconds=int(now.strftime('%f')))

In [4]: now
Out[4]: datetime.datetime(2020, 7, 16, 16, 31, 41, 184395)

In [5]: utc_now
Out[5]: datetime.datetime(2020, 7, 16, 20, 31, 41, 184395)
```

Looks a little bit ugly, I agree.

samdoran committed:
Use time.time() for getting the current time

Convert from that stored epoch timestamp to local and UTC times

sivel commented on lib/ansible/module_utils/facts/system/date_time.py:
```diff
@@ -18,9 +18,10 @@
-import datetime
 import time
 
+from datetime import datetime
```
I'd be cautious doing this. Because the class is the same name as the module, this often trips people up.  I know it's  longer, but I just rely on using `datetime.datetime.` everywhere, to avoid the potential confusion that this import can create.

samdoran replied:
That's fine. I think this also made the patching a bit cleaner because I could patch the `datetime` imported in `ansible.module_utils.facts.system.date_time`, but I can change it.

nickgryg commented on test/units/module_utils/facts/test_date_time.py:
```diff
@@ -7,29 +7,45 @@
 def fake_now(monkeypatch):
-    """Patch `datetime.datetime.now()` to return a deterministic value."""
+    """
+    Patch `datetime.datetime.fromtimestamp()`, `datetime.datetime.utcfromtimestamp()`,
```
I think we don't need to patch `datetime.datetime.fromtimestamp()` and `datetime.datetime.utcfromtimestamp()` methods as they just return a datetime object from patched `time.time()`.

webknjaz replied:
+1 it's better to just patch `time.time`

samdoran replied:
I tried not patching those but ended up incorrect `datetime` objects. For example:
```
>       assert date_facts['date_time'][fact_name] == fact_value
E       AssertionError: assert '2020-07-11T06:34:56.123456Z' == '2020-07-11T02:34:56.124356Z'
E         - 2020-07-11T02:34:56.124356Z
E         ?             ^          -
E         + 2020-07-11T06:34:56.123456Z
E         ?             ^         +

```

```
================================================ short test summary info =================================================
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[hour-12]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[epoch-1594434896]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[time-12:34:56]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601_basic-20200711T123456124356]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601_basic_short-20200711T123456]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601_micro-2020-07-11T02:34:56.124356Z]
FAILED test/units/module_utils/facts/test_date_time.py::test_date_time_facts[iso8601-2020-07-11T02:34:56Z]
```

webknjaz replied:
@samdoran then, I'd say that there's something wrong with the setup and TZ should be adjusted somewhere.

samdoran replied:
That's what I was thinking. I tried moving the application of the TZ around, but didn't have any luck. I'll keep working on it.

webknjaz replied:
So I looked it up and it seems that it hits `time.localtime()` which internals are buried in C:

https://github.com/python/cpython/blob/0275e04/Lib/datetime.py#L829 -> https://github.com/python/cpython/blob/0275e04/Modules/timemodule.c#L497 -> https://github.com/python/cpython/blob/0275e04/Python/pytime.c#L1053 -> https://en.cppreference.com/w/c/chrono/localtime.

It says that `tzset()` will set up internal structures based on `TZ` env var but this seems dangerous and you risk poisoning the entire interpreter state, hence influence other tests if you don't figure out how to clean it up properly.

Looks like it's really safer to monkey-patch methods in the Python-land...

webknjaz commented on lib/ansible/module_utils/facts/system/date_time.py:
```diff
@@ -48,8 +53,8 @@ def collect(self, module=None, collected_facts=None):
             date_time_facts['epoch'] = str(int(time.time()))
```
```suggestion
            date_time_facts['epoch'] = str(int(epoch_ts))
```

samdoran replied:
Nice catch.

samdoran committed:
Used existing timestamp for epoch time

nickgryg commented on test/units/module_utils/facts/test_date_time.py:
```diff
@@ -7,29 +7,45 @@
+from ansible.module_utils.facts.system import date_time
 
-TIMESTAMP = datetime.datetime(2020, 7, 11, 12, 34, 56, 124356)
+EPOCH_TS = 1594449296.123456
```
```suggestion
EPOCH_TS = 1594449296.124356
```

samdoran replied:
I'm not sure I understand why this change is necessary. Could you enlighten me? 😄

sivel replied:
Looks like values below in `DT` and `DT_UTC` are 124356, instead of 123456.

nickgryg replied:
Originally, there was `124356` microseconds in unit tests, instead of `123456`. At least two tests are rely on this figure.

samdoran committed:
Change datetime import

samdoran committed:
Do not set timezone with time.tzset()

Since it changes the test environment, it's not a good idea.

samdoran committed:
Add tests for problematic date_time facts

Don't attempt to modify the timezone, but instead verify
the call args when possible and the return value.

samdoran committed:
Validate epoch length, use sets for checking letters

samdoran committed:
Don't bother spying on the calls

samdoran committed:
Docs correction

samdoran commented on test/units/module_utils/facts/test_date_time.py:
```diff
@@ -0,0 +1,114 @@
+    assert not set(fake_date_facts['date_time']['tz']).difference(set(string.ascii_uppercase))
+
+
+def test_date_time_tz_dst(fake_date_facts):
```
Looks like this can now be combined with the previous test since I'm no longer spying on the calls.

samdoran committed:
Combine timezone tests
